python/ray/experimental/workflow/tests/test_virtual_actor_2.py

```diff
@@ -89,6 +89,254 @@ def test_indirect_actor_writer(workflow_start_regular):
+            self._counter = v
+
+    locks = [FileLock(f) for f in file_lock]
+    for lock in locks:
+        lock.acquire()
+
+    c = Counter.get_or_create("counter")
+    ray.get(c.ready())
+    final_ret = c.incr.run_async(len(file_lock) - 1)
+    for i in range(0, len(file_lock) - 2):
+        locks[-i - 1].release()
+        val = c.val.run()
+        for _ in range(0, 60):
+            if val == i + 1:
+                break
+            val = c.val.run()
+            time.sleep(1)
+        assert val == i + 1
+
+    fail_flag.touch()
+    locks[1 - len(file_lock)].release()
+    # Fail the pipeline
+    with pytest.raises(Exception):
+        ray.get(final_ret)
+
+    fail_flag.unlink()
+    workflow.resume("counter")
+    # After resume, it'll start form the place where it failed
+    for i in range(len(file_lock) - 1, len(file_lock)):
+        locks[-i - 1].release()
+        val = c.val.run()
+        for _ in range(0, 60):
+            if val == i + 1:
+                break
+            val = c.val.run()
+            time.sleep(1)
+        assert val == i + 1
+
+    assert c.val.run() == 5
+
+    c.write.run(10)
+    assert "\n".join(
+        [str(i) for i in range(10)]) == (tmp_path / "record").read_text()
+
+
+@pytest.mark.parametrize(
+    "workflow_start_regular",
+    [{
+        "num_cpus": 4
+        # We need more CPUs, otherwise 'create()' blocks 'get()'
+    }],
+    indirect=True)
+@pytest.mark.skip(
+    reason="Return a list of virtual actor sub method is not supported.")
+def test_wf_in_actor_seq(workflow_start_regular, tmp_path):
+    record = tmp_path / "record"
+    record.touch()
+
+    @workflow.step
+    def join(*args):
+        return args
+
+    @workflow.virtual_actor
+    class Q:
+        def write(self, n):
+            r = record.read_text()
+            r += f"{n}\n"
+            record.write_text(r)
+
+        def write_queue(self, n):
+            return join.step(*[self.write.step(i) for i in range(n)])
+
+        def __getstate__(self):
+            return None
+
+        def __setstate__(self, v):
+            pass
+
+    q = Q.get_or_create("queue")
+    q.write_queue.run(5)
+    assert "\n".join([str(i) for i in range(5)]) == record.read_text()
+
+
+@pytest.mark.parametrize(
+    "workflow_start_regular",
+    [{
+        "num_cpus": 4
+        # We need more CPUs, otherwise 'create()' blocks 'get()'
+    }],
+    indirect=True)
+def test_wf_in_actor_seq_2(workflow_start_regular, tmp_path):
+    record = tmp_path / "record"
+    record.touch()
+
+    file_lock = tmp_path / "lock"
+
+    @workflow.step
+    def join(*args):
+        return args
+
+    @workflow.step
+    def write(n):
+        with FileLock(str(file_lock)):
+            r = record.read_text()
+            r += f"{n}\n"
+            record.write_text(r)
+
+    @workflow.virtual_actor
+    class Q:
+        def write_queue(self, n):
+            return join.step(*[write.step(i) for i in range(n)])
```

Comment:
@ericl  This will work, because write is not a virtual actor method, so it's not like virtual actor workflow inside a normal workflow.